shared/src/pool_fetching.rs

```diff
@@ -20,30 +21,38 @@ type RelativeReserves = (U256, U256, H160);
 
 #[async_trait::async_trait]
 pub trait PoolFetching: Send + Sync {
-    async fn fetch(&self, token_pairs: HashSet<TokenPair>) -> Vec<Pool>;
+    async fn fetch(&self, token_pairs: HashSet<TokenPair>) -> Result<Vec<Pool>>;
 }
 
 pub struct FilteredPoolFetcher {
     inner: Box<dyn PoolFetching>,
-    unsupported_tokens: HashSet<H160>,
+    bad_token_detector: Arc<dyn BadTokenDetecting>,
 }
 
 impl FilteredPoolFetcher {
-    pub fn new(inner: Box<dyn PoolFetching>, unsupported_tokens: HashSet<H160>) -> Self {
+    pub fn new(
+        inner: Box<dyn PoolFetching>,
+        bad_token_detector: Arc<dyn BadTokenDetecting>,
+    ) -> Self {
         Self {
             inner,
-            unsupported_tokens,
+            bad_token_detector,
         }
     }
 }
 
 #[async_trait::async_trait]
 impl PoolFetching for FilteredPoolFetcher {
-    async fn fetch(&self, token_pairs: HashSet<TokenPair>) -> Vec<Pool> {
-        let filtered_pairs = token_pairs
-            .into_iter()
-            .filter(|pair| !self.unsupported_tokens.iter().any(|t| pair.contains(t)))
-            .collect();
+    async fn fetch(&self, pairs: HashSet<TokenPair>) -> Result<Vec<Pool>> {
+        let mut filtered_pairs = HashSet::new();
+        'outer: for pair in pairs {
```

Comment:
This is new for me `outer!

Reply:
Not sure if it's in the book but I found this documentation https://doc.rust-lang.org/rust-by-example/flow_control/loop/nested.html .